scripts/firmware.py

```diff
@@ -0,0 +1,2097 @@
+import os
+import re
+import subprocess
+import argparse
+import shutil
+import tempfile
+import fnmatch
+import platform
+import time
+
+from enum import Enum  # type: ignore
+from time import sleep
+from typing import Optional, Tuple, Callable, NamedTuple, Dict, List, Union
+
+HOME_PATH = os.path.expanduser("~")
+CONFIG_DIR: str = os.path.expanduser("~/printer_data/config")
+KLIPPY_LOG: str = os.path.expanduser("~/printer_data/logs/klippy.log")
+KLIPPER_DIR: str = os.path.expanduser("~/klipper")
+KATAPULT_DIR: str = os.path.expanduser("~/katapult")
+
+FLASHER_VERSION: str = "0.0.1"
+
+PAGE_WIDTH: int = 89  # Default global width
+
+is_advanced: bool = False
+
+
+class Color(Enum):
+    RESET = "\033[0m"
+    RED = "\033[91m"
+    GREEN = "\033[92m"
+    YELLOW = "\033[93m"
+    MAGENTA = "\033[95m"
+    CYAN = "\033[96m"
+
+
+# Define a custom namespace class
+class FirmwareNamespace(argparse.Namespace):
+    branch: str = "master"
+    debug: bool = False
+    type: bool = False
+    high_temp: bool = False
+    all: bool = False
+    kseries: bool = False
+    device: Optional[str] = None
+    flash: Optional[str] = None
+
+
+class Version(NamedTuple):
+    parts: Tuple[int, ...]
+
+    @classmethod
+    def from_string(cls, version: str) -> "Version":
+        return cls(tuple(int(part) for part in version.split(".")))
+
+
+class FirmwareFile(NamedTuple):
+    subdirectory: str
+    filename: str

```

Comment:
Or `TypedDict`, depends if you want a tuple or a dict.

Reply:
Should be fine as a tuple, Doesnt need to be mutable I dont think at this time.